x/concentrated-liquidity/simulation/sim_msgs.go

```diff
@@ -61,6 +64,211 @@ func RandomMsgCreateConcentratedPool(k clkeeper.Keeper, sim *osmosimtypes.SimCtx
+	}
+
+	if len(positions) == 0 {
+		return nil, fmt.Errorf("user does not have any position")
+	}
+
+	// pick a random position
+	randPosition := positions[rand.Intn(len(positions))]
+
+	return &cltypes.MsgCollectFees{
+		PoolId:    randPosition.PoolId,
+		Sender:    sender.Address.String(),
+		LowerTick: randPosition.LowerTick,
+		UpperTick: randPosition.UpperTick,
+	}, nil
+}
+
+// RandomSwapExactAmountIn utilizes a random pool and swaps and exact amount in for minimum of the secondary pool token
+func RandomCLSwapExactAmountIn(k clkeeper.Keeper, sim *osmosimtypes.SimCtx, ctx sdk.Context) (*gammtypes.MsgSwapExactAmountIn, error) {
+	// get random pool, randomly select one of the pool denoms to be the coinIn, other is coinOut
+	clPool, poolDenoms, err := getRandCLPool(k, sim, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// set the swap route to use this pool
+	route := []poolmanagertypes.SwapAmountInRoute{{
+		PoolId:        clPool.GetId(),
+		TokenOutDenom: poolDenoms[1],
+	}}
+
+	// find an address that has a balance of the coinIn
+	sender, accCoinIn, senderExists := sim.SelAddrWithDenom(ctx, poolDenoms[0])
+	if !senderExists {
+		return nil, fmt.Errorf("no sender with denom %s exists", poolDenoms[0])
+	}
+
+	// select a random amount that is upper-bound by the address's balance of coinIn
+	randomCoinSubset := sim.RandSubsetCoins(sdk.NewCoins(sdk.NewCoin(accCoinIn.Denom, accCoinIn.Amount)))
+
+	poolI := clPool.(poolmanagertypes.PoolI)
+
+	// calculate the minimum number of tokens received from input of tokenIn
+	tokenOutMin, err := k.CalcOutAmtGivenIn(ctx, poolI, randomCoinSubset[0], poolDenoms[1], clPool.GetSwapFee(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	return &gammtypes.MsgSwapExactAmountIn{
+		Sender:            sender.Address.String(),
+		Routes:            route,
+		TokenIn:           randomCoinSubset[0],
+		TokenOutMinAmount: tokenOutMin.Amount,
+	}, nil
+}
+
+// RandomSwapExactAmountOut utilizes a random pool and swaps a max amount amount in for an exact amount of the secondary pool token
+func RandomCLSwapExactAmountOut(k clkeeper.Keeper, sim *osmosimtypes.SimCtx, ctx sdk.Context) (*gammtypes.MsgSwapExactAmountOut, error) {
+	// get random pool, randomly select one of the pool denoms to be the coinIn, other is coinOut
+	clPool, poolDenoms, err := getRandCLPool(k, sim, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// set the swap route to use this pool
+	route := []poolmanagertypes.SwapAmountOutRoute{{
+		PoolId:       clPool.GetId(),
+		TokenInDenom: poolDenoms[0],
+	}}
+
+	// find an address that has a balance of the coinIn
+	senderAcc, accCoin, senderExists := sim.SelAddrWithDenom(ctx, poolDenoms[0])
+	if !senderExists {
+		return nil, fmt.Errorf("no sender with denom %s exists", poolDenoms[0])
+	}
+
+	// set the subset of coins to be upper-bound to the minimum between the address and the pool itself
+	randomCoinSubset := sim.RandSubsetCoins(sdk.NewCoins(sdk.NewCoin(accCoin.Denom, accCoin.Amount)))
```

Comment:
can i please get help on this? Currently, `RandomCLSwapExactAmountOut` fails because i believe it tries to `CalcOutAmtGivenIn` the sender coin instead of  min(senderAmountIn, poolAmountIn) which is what we do in gamm([example](https://github.com/osmosis-labs/osmosis/blob/main/x/gamm/simulation/sim_msgs.go#L184)) 

I dont think there is a way we can calculate poolAssetAmounts for CL pools from our previous conversation. If anyone has any feedback on this i'd really appreciate :))

Reply:
I tried calculating liquidity using these equations but not sure if we have enough information 

```
def calc_liquidity_0(amount_zero: sp.Float, sqrtPriceA: sp.Float, sqrtPriceB: sp.Float) -> sp.Float:
    """Calculates and returns liquidity zero. 
    """
    return amount_zero * (sqrtPriceA * sqrtPriceB) / (sqrtPriceB - sqrtPriceA)

def calc_liquidity_1(amount_one: sp.Float, sqrtPriceA: sp.Float, sqrtPriceB: sp.Float) -> sp.Float:
    """Calculates and returns liquidity one. 
    """
    return amount_one * (sqrtPriceB - sqrtPriceA)

```

Reply:
Im not sure if I understood the question correctly, what is it that you are trying to achieve? We only need tokenOut and swapfees specified for the swapExactAmountOut don't we?

Reply:
We also need tokenInMaxAmount and thats where the issue is 
https://github.com/osmosis-labs/osmosis/pull/4241/files#diff-f01e8795f0fb3ac50fd61ff5441aac0f1e4153629f16c05da9c3a8785c408dc9R264-R268